### PR TITLE
Add 'crop' option to RandomRotation

### PR DESCRIPTION
**Pull request by sg3-141-592**
Add 'crop' option to `RandomRotation` that centre crops the rotated image to remove any padding regions introduced by the rotation #9147 

image

`crop` and `expand` are mutually exlusive.

This is based off the approach in [albumentations - rotate.py](https://github.com/albumentations-team/albumentations/blob/main/albumentations/augmentations/geometric/rotate.py). Which in turn is taken from here https://stackoverflow.com/questions/16702966/rotate-image-and-crop-out-black-borders. Slight difference is we use int() (floor) for the final crop dimensions rather than Albumentations' coordinate-based rounding, to guarantee no fill pixels appear in the output.

I've added a `RandomRotation._extract_params_for_v1_transform` override to strip the crop parameter for `torch.jit.script()` compatibility, and to raise a `ValueError` if `crop=True` since the v1 transform has no equivalent.

### CODE REVIEW

File: torchvision/transforms/v2/functional/_geometry.py [hunk 1]
```diff
@@ -1343,6 +1343,48 @@ def affine_video(
...
+            crop_w, crop_h = x / cos_a, x / sin_a
+    else:
+        # Fully constrained: crop touches all four sides
+        cos_2a = cos_a * cos_a - sin_a * sin_a
+        crop_w = (width * cos_a - height * sin_a) / cos_2a
+        crop_h = (height * cos_a - width * sin_a) / cos_2a
+
+    # Use floor (int()) to guarantee the crop region contains no fill pixels.
+    # Clamp to image dimensions for edge cases like wide images rotated near 90°.
+    return min(int(crop_h), height), min(int(crop_w), width)
```
Comment (zy1git) [inline torchvision/transforms/v2/functional/_geometry.py:1385]: Would math.floor() be more appropriate here? int() truncates toward zero rather than rounding down, and math.floor() would make the "round down" intent more explicit. (Both behave the same for positive values, which is always the case here, so this is just a readability suggestion.)

File: test/test_transforms_v2.py [hunk 2]
```diff
@@ -2462,6 +2462,62 @@ def test_functional_image_fast_path_correctness(self, size, angle, expand):
...
+    @pytest.mark.parametrize("size", [(100, 100), (120, 80)])
+    @pytest.mark.parametrize("angle", [15.0, 30.0, 45.0])
+    def test_transform_crop_removes_fill(self, size, angle):
+        # Output of crop=True should contain no fill pixels when input is fully non-zero
+        h, w = size
+        image = tv_tensors.Image(torch.full((3, h, w), 200, dtype=torch.uint8))
+        transform = transforms.RandomRotation((angle, angle), fill=0, crop=True)
+        output = transform(image)
+        assert output.min().item() > 0, "crop=True output should have no fill pixels"
+        assert output.shape[-2] < h or output.shape[-1] < w, "crop=True should reduce at least one dimension"
```
Comment (zy1git) [inline test/test_transforms_v2.py:2474]: We don't need specify anything beyond the assert, pytest is already good at showing the right thing. Please remove the message after assert.

File: test/test_transforms_v2.py [hunk 3]
```diff
@@ -2462,6 +2462,62 @@ def test_functional_image_fast_path_correctness(self, size, angle, expand):
...
+        out_image, out_mask, out_boxes = transform(image, mask, boxes)
+        assert out_image.shape[-2:] == out_mask.shape[-2:]
+        assert out_boxes.canvas_size == (out_image.shape[-2], out_image.shape[-1])
+
+    def test_transform_crop_and_expand_mutually_exclusive(self):
+        with pytest.raises(ValueError, match="crop and expand are mutually exclusive"):
+            transforms.RandomRotation(30, expand=True, crop=True)
+
+    @pytest.mark.parametrize("angle", [0.0, 90.0, 180.0, 270.0])
+    def test_transform_crop_zero_angle_preserves_size(self, angle):
```
Comment (zy1git) [inline test/test_transforms_v2.py:2498]: The test name "test_transform_crop_zero_angle_preserves_size" says "zero_angle" but it tests 0°, 90°, 180°, and 270°, which is a bit misleading. Could you please rename it?

File: test/test_transforms_v2.py [hunk 4]
```diff
@@ -2462,6 +2462,62 @@ def test_functional_image_fast_path_correctness(self, size, angle, expand):
...
+        # Multiples of 90° should not reduce the image size
+        image = tv_tensors.Image(torch.zeros(3, 100, 100, dtype=torch.uint8))
+        transform = transforms.RandomRotation((angle, angle), crop=True)
+        output = transform(image)
+        assert output.shape == image.shape
+
+    def test_largest_inscribed_crop_size(self):
+        from torchvision.transforms.v2.functional._geometry import _largest_inscribed_crop_size
```
Comment (zy1git) [inline test/test_transforms_v2.py:2506]: We import the functions on the top of the file. For consistency, please move the `_largest_inscribed_crop_size` import to the top of the file too.

File: test/test_transforms_v2.py [hunk 5]
```diff
@@ -2462,6 +2462,62 @@ def test_functional_image_fast_path_correctness(self, size, angle, expand):
...
+
+        # No rotation: crop equals original size
```
Comment (zy1git) [inline test/test_transforms_v2.py:2508]: This comment is obvious from the code, so we can remove it.

File: test/test_transforms_v2.py (hunk 3 above)
Comment (zy1git) [inline test/test_transforms_v2.py:2498]: Can we add non-square sizes by parametrizing over non-square sizes like (120, 80) here?

**Review by zy1git — COMMENTED** [review]
@sg3-141-592 Thanks a lot for this PR! I left some review comments. Feel free to take a look and address them. 

By the way, there are no tests for negative angles. Can we add one? Something like `asserting _largest_inscribed_crop_size(100, 100, -30) == _largest_inscribed_crop_size(100, 100, 30)` to verify the `abs()`on sin/cos handles negative angles correctly.

Also, please address the test failure for linting. We need to format files with µfmt.

File: test/test_transforms_v2.py (hunk 4 above)
Comment (sg3-141-592) [reply test/test_transforms_v2.py]: Moved to top of file with the rest of the imports ✔️

File: test/test_transforms_v2.py (hunk 5 above)
Comment (sg3-141-592) [reply test/test_transforms_v2.py]: Removed ✔️

File: test/test_transforms_v2.py (hunk 3 above)
Comment (sg3-141-592) [reply test/test_transforms_v2.py]: Sure, updated to `test_transform_crop_right_angle_preserves_size`

File: torchvision/transforms/v2/functional/_geometry.py (hunk 1 above)
Comment (sg3-141-592) [reply torchvision/transforms/v2/functional/_geometry.py]: I agree, `math.floor` is more clear. Updated ✔️

File: test/test_transforms_v2.py (hunk 2 above)
Comment (sg3-141-592) [reply test/test_transforms_v2.py]: Sure, removed the superfluous messages

File: test/test_transforms_v2.py (hunk 3 above)
Comment (sg3-141-592) [reply test/test_transforms_v2.py]: Now replaced with two test functions because of an issue found below with different canvas sizes for this test case
- `test_transform_crop_half_turn`
- `test_transform_crop_quarter_turn`

Comment (sg3-141-592) [reply test/test_transforms_v2.py]: Sure, adding this test case found an issue with our early returns for right angle rotations. For 90 deg and 270 deg we need to return the `min(w,h)` square that sits inside the canvas.

``` python
def _largest_inscribed_crop_size(width: int, height: int, angle: float):
...
    # Guard against small values of sin_a or cos_a that will cause us to truncate a pixel
    # off rotations that introduce no padding (e.g. 90° or 180°).
    if sin_a < 1e-10:
        return height, width
    if cos_a < 1e-10:
-       return width, height
+       # 90°/270°: rotated rectangle has dims swapped; clamp to canvas (h, w).
+      return min(width, height), min(height, width)
```

image

Note: With this fix I've decided to split `test_transform_crop_right_angle_preserves_size` into two different cases for the different early exit paths
- `test_transform_crop_half_turn`
- `test_transform_crop_quarter_turn`

**Comment by sg3-141-592** [comment]
Added test suggested test case for negative angles

``` python
def test_largest_inscribed_crop_size(self):
...
         # Negative angles are symmetric to positive ones (abs() on sin/cos)
        for w, h, a in [(100, 100, 30), (200, 100, 15), (120, 80, 45)]:
            assert _largest_inscribed_crop_size(w, h, -a) == _largest_inscribed_crop_size(w, h, a)
```

**Comment by zy1git** [comment]
@sg3-141-592 Thanks a lot for addressing the comments in detail!

@NicolasHug Could you please take a look at this PR as well? Please feel free to leave any additional comments.
